### PR TITLE
Alternative extractor implementation

### LINKED ISSUE #92: Support ML-style active patterns
See the documentation about "active patterns" in [MLStyle](https://thautwarm.github.io/MLStyle.jl/latest/syntax/pattern.html#active-patterns) and [F#](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns).  We should consider supporting those.

### PR DESCRIPTION
**Pull request by nystrom**
This is an alternative implementation of extractors that uses `Type` rather than `Val{<:Symbol}` as in #114 for the dispatch. If the `extract` method is defined for the given type, the extractor pattern is used; otherwise, the struct pattern is used.

Fixes #92

### CODE REVIEW

File: src/binding.jl [hunk 1]
```diff
@@ -242,52 +260,102 @@ function bind_pattern!(
...
-                    error("$(location.file):$(location.line): Type `$bound_type` has " *
-                          "no field `$field_name`.")
-                end
+        # If the extractor failed, try the field-by-field match.
+        type_conjuncts = BoundPattern[]
+
+        if isnothing(extractor_temp)
+            # Ensure that the extractor actually failed.
+            # Otherwise, we could have an ambiguity between the type pattern and the extractor pattern.
+            if !isnothing(extractor_temp)
```
Comment (gafter) [inline src/binding.jl:295]: Given the test at line 292, I don't see how this can happen.

File: src/binding.jl [hunk 2]
```diff
@@ -242,52 +260,102 @@ function bind_pattern!(
...
             end
 
-            field_type = nothing
-            if field_name == match_fieldnames(Symbol)[1]
-                # special case Symbol's hypothetical name field.
-                field_type = String
+            field_names::Tuple = match_fieldnames(bound_type)
+
+            if match_positionally && len != length(field_names)
+                # If the extractor is defined, silently fail if the field-by-field match fails.
```
Comment (gafter) [inline src/binding.jl:303]: These are strange semantics. Why wouldn't you use the extractor if the type has the wrong arity?

File: src/binding.jl [hunk 3]
```diff
@@ -242,52 +260,102 @@ function bind_pattern!(
...
+                if isnothing(extractor_temp)
+                    error("$(location.file):$(location.line): The type `$bound_type` has " *
+                            "$(length(field_names)) fields but the pattern expects $len fields.")
+                else
+                    pattern0 = BoundFalsePattern(location, source)
```
Comment (gafter) [inline src/binding.jl:308]: There is not test for this branch.

File: src/pretty.jl [hunk 4]
```diff
@@ -185,6 +185,13 @@ function pretty(io::IO, p::BoundFetchLengthPattern)
     pretty(io, p.input)
     print(io, ")")
 end
+function pretty(io::IO, p::BoundFetchExtractorPattern)
```
Comment (gafter) [inline src/pretty.jl:188]: Please exercise this code in a test.

File: src/binding.jl [hunk 5]
```diff
@@ -66,6 +66,21 @@ function bind_type(location, T, input, binder)
 
     bound_type
 end
+function try_bind_type(location, T, input, binder)
```
Comment (gafter) [inline src/binding.jl:69]: This appears to be dead code.

File: src/binding.jl (hunk 2 above)
Comment (gafter) [reply src/binding.jl:303]: Or, if you're not going to use the extractor, give the diagnostic.

File: src/binding.jl (hunk 1 above)
Comment (nystrom) [reply src/binding.jl]: This was WIP. Fixed.

File: src/pretty.jl (hunk 4 above)
Comment (nystrom) [reply src/pretty.jl:188]: done

**Comment by gafter** [comment]
I love the idea of using the type instead of a token, for exactly the reason you mention. I will look at this (from scratch) today or tomorrow.

**Comment by gafter** [comment]
One thing occurs to me. Extractors in this version pay no attention to the arity of the pattern. So, for example, the following would fail
```
    AddExpr(_, _, _) => true
```
even though the struct type has three fields, because the extractor produces two values, not three.

Is this a problem? Should the extractor specify the arity of the tuple it will produce? If so, would that interfere with the use of named tuples?

**Comment by nystrom** [comment]
Right, one issue with extractors is that you lose all the nice error checking on the subpatterns. I was thinking you could pass in the expected arity to the extractor:
```
# called for pattern AddExpr(x, y)
function extract(::Type{AddExpr}, ::Val{2}, v)
   return (v.left, v.right,)
end
```
But, I didn't want to complicate things too much.

In the case of named arguments, you could pass in the desired field names (as `Val`):
```
# called for pattern AddExpr(left=x, right=y)
function extract(::Type{AddExpr}, ::Val{:left}, ::Val{::right}, v)
    return (; left = v.left, right = v.right)
end
```
I fear this is getting much too complicated, however.

**Comment by nystrom** [comment]
@gafter I've implemented arity support.

**Comment by gafter** [comment]
My point about the named tuples is that if you have the function specify the arity, then we can handle things like `Foo(1, 2, 3)` nicely (using the type pattern because the user-specified pattern doesn't match the arity). However, if we extend type patterns to handle "keyword argument"-like things like `Foo(1, 2; y=7)`, we can't easily extend this scheme to handle that very well (what is its arity? two? three? what should the function return?).

I'm probably overthinking this.

**Comment by nystrom** [comment]
I'm not sure how `Foo(1, 2; y=7)` works when `Foo` is just a struct. Would you look for a constructor?
I suppose you could make it equivalent to `Foo(1,2) && Foo(y=7)`.

File: src/binding.jl (hunk 2 above)
Comment (gafter) [reply src/binding.jl:279]: Or, perhaps delete this line comment that is confusing me.

**Review by gafter — COMMENTED** [review]
Please bump the version number to `2.3.0` (in `Project.toml`) as this adds a feature.

**Review by gafter — APPROVED** [review]
Looks great!

**Pull request merged**